components/event-publisher/event-output-adapters/org.wso2.carbon.event.output.adapter.email/src/main/java/org/wso2/carbon/event/output/adapter/email/EmailEventAdapter.java

```diff
@@ -356,4 +379,52 @@ public void run() {
 
     }
 
+    private void LogMessagingException(MessagingException e, String mailRecipient) {
+
+        if (e instanceof SendFailedException) {
+            List<String> sentMails = new ArrayList<>();
+            List<String> invalidMails = new ArrayList<>();
+            List<String> unsentMails = new ArrayList<>();
+
+            Address[] sent = ((SendFailedException) e).getValidSentAddresses();
+            if (ArrayUtils.isNotEmpty(sent)) {
+                for (Address address : sent) {
+                    sentMails.add(address.toString());
+                }
+            }
+
+            Address[] invalid = ((SendFailedException) e).getInvalidAddresses();
+            if (ArrayUtils.isNotEmpty(invalid)) {
+                for (Address address : invalid) {
+                    invalidMails.add(address.toString());
+                }
+            }
+
+            Address[] unsent = ((SendFailedException) e).getValidUnsentAddresses();
+            if (ArrayUtils.isNotEmpty(unsent)) {
+                for (Address address : unsent) {
+                    unsentMails.add(address.toString());
+                }
+            }
+
+            log.error(String.format("Exception occurred when sending email to %s. %s. Sent email=%s, " +
+                            "Invalid emails=%s, Unsent valid emails=%s",
+                    mailRecipient,
+                    e.getMessage(),
+                    StringUtils.join(sentMails, ","),
+                    StringUtils.join(invalidMails, ","),
+                    StringUtils.join(unsentMails, ",")), e);
+
+        }
+
+        log.error("Exception occurred when sending email to " + mailRecipient + ". " + e.getMessage(), e);
+
+        // MessagingException has the capability to chain exceptions.
+        // Therefore checking for any chained exceptions.
+        Exception nextEx = e.getNextException();
+        if (nextEx instanceof MessagingException ) {
```

Comment:
Need to avoid the rare possibility of StackOverflow exception due to this.

Reply:
fixed